File: src/lexer/tokens.go
package lexer

type TokenType int

const (
	IDENTIFIER TokenType = iota

	// separators
	SEMI_COLON
	LEFT_PAREN
	RIGHT_PAREN
	LEFT_BRACE
	RIGHT_BRACE

	// operators
	EQUAL
	PLUS
	MINUS
	PLUS_PLUS
	LESS
	MULTIPLY
	DIVIDE

	// literals
	STRING
	NUMBER

	KEYWORD
)

type Token struct {
	Value string
	Type  TokenType
}

func (t Token) String() string {

	if t.Type == STRING || t.Type == IDENTIFIER || t.Type == NUMBER {
		return "Type: " + TokenTypeString(t.Type) + " Value: '" + t.Value + "'"
	}

	return "Type: " + TokenTypeString(t.Type)
}

func TokenTypeString(t TokenType) string {
	switch t {
	case IDENTIFIER:
		return "identifier"
	case KEYWORD:
		return "keyword"
	case EQUAL:
		return "equal"
	case STRING:
		return "string"
	case SEMI_COLON:
		return "semi_colon"
	case NUMBER:
		return "number"
	case PLUS:
		return "plus"
	case MINUS:
		return "minus"
	case MULTIPLY:
		return "multiply"
	case DIVIDE:
		return "divide"
	case PLUS_PLUS:
		return "plus_plus"
	case LESS:
		return "less"
	case LEFT_PAREN:
		return "left_paren"
	case RIGHT_PAREN:
		return "right_paren"
	case LEFT_BRACE:
		return "left_brace"
	case RIGHT_BRACE:
		return "right_brace"
	default:
		return "unknown"
	}
}

File: src-ts/tokens.ts
export enum TokenType {
  IDENTIFIER = "identifier",

  // separators
  SEMI_COLON = ";",
  LEFT_PAREN = "(",
  RIGHT_PAREN = ")",
  LEFT_BRACE = "[",
  RIGHT_BRACE = "]",
  LEFT_CURLY = "{",
  RIGHT_CURLY = "}",

  // operators
  PLUS_PLUS = "++",
  EQUAL = "=",
  PLUS = "+",
  MINUS = "-",
  LESS = "<",
  MULTIPLY = "*",
  DIVIDE = "/",

  // literals
  STRING = "string",
  NUMBER = "number",

  KEYWORD = "keyword",
}

export type Token = {
  value: string;
  type: TokenType;
};

export function is_keyword(value: string): boolean {
  return ["var", "for"].includes(value);
}
